api/client/lib/authn.go

```diff
@@ -0,0 +1,211 @@
+package lib
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// authnOpts is the set of authentication-related options which the client
+// should be using.
+type authnOpts struct {
+	getBasic  func(realm string) (string, string, error)
+	getBearer func(challenge string) (string, error)
+	negotiate bool
+}
+
+// SetBasicAuth sets a callback which the client object can use to obtain
+// credentials if it needs to perform Basic authentication to the daemon.
+func (cli *Client) SetBasicAuth(get func(challenge string) (string, string, error)) {
+	cli.authnOpts.getBasic = get
+}
+
+// SetBearerAuth sets a callback which the client object can use to obtain
+// credentials if it needs to perform Basic authentication to the daemon.
+func (cli *Client) SetBearerAuth(get func(challenge string) (string, error)) {
+	cli.authnOpts.getBearer = get
+}
+
+// SetNegotiateAuth controls whether or not the client is allowed to attempt to
+// use Negotiate authentication.
+func (cli *Client) SetNegotiateAuth(enabled bool) {
+	cli.authnOpts.negotiate = enabled
+}
+
+// authResponder is an interface that wraps the scheme,
+// authRespond, and authCompleted methods.
+//
+// At initialization time, an implementation of authResponder should register
+// itself by calling registerAuthResponder.
+type authResponder interface {
+	// Scheme should return the name of the authorization scheme for which
+	// the responder should be called.
+	scheme() string
+	// authRespond, given the authentication header value associated with
+	// the scheme that it implements, can decide if the request should be
+	// retried.  If it returns true, then the request is retransmitted to
+	// the server, presumably because it has added an authentication header
+	// which it believes the server will accept.
+	authRespond(cli *Client, challenge string, req *http.Request) (bool, error)
+	// AuthCompleted, given a (possibly empty) WWW-Authenticate header from
+	// a successful response, should decide if the server's reply should be
+	// accepted.
+	authCompleted(cli *Client, challenge string, req *http.Request) (bool, error)
+}
+
+// authResponderCreator either creates a new authResponder, or returns nil
+type authResponderCreator func() authResponder
+
+var authResponderCreators = []authResponderCreator{}
```

Comment:
If we switch from storing arrays in the map to single objects, the main purpose of using functions to create objects rather than using singletons is to let authentication implementations keep their state in the Client object rather than making it global - I'm thinking mainly of caching the creds for Basic, and partially-established contexts in Negotiate.